Add renode-test integration

mateusz-holenko opened the pull request:
This PR adds a cmake `run_renode_test` target and integrates it with twister using a new `renode_test` harness.

It allows to express new test cases in Robot Framework syntax and run them in Renode.
As an example we provide a robot file for the `shell` sample and extend the `yaml` file accordingly.

To run a selected robot test in `twister` execute:
```
./scripts/twister --platform hifive1 --test samples/subsys/shell/shell_module/sample.shell.shell_module.robot -v
```

It's also possible to run it outside `twister`, with:
```
ROBOT_FILES=shell_module.robot west build -p -b hifive1 -s samples/subsys/shell/shell_module -t run_renode_test
```

mszprejda committed:
m2gl025_miv: Add RENODE_UART variable to board.cmake

Set RENODE_UART variable describing which UART device
should be used when running Robot Framework test on the platform.

Signed-off-by: Michał Szprejda <mszprejda@antmicro.com>
Signed-off-by: Mateusz Hołenko <mholenko@antmicro.com>

mszprejda committed:
hifive1: Add RENODE_UART variable to board.cmake

Set RENODE_UART variable describing which UART device
should be used when running Robot Framework test on the platform.

Signed-off-by: Michał Szprejda <mszprejda@antmicro.com>
Signed-off-by: Mateusz Hołenko <mholenko@antmicro.com>

mszprejda committed:
shell_module: Add a robot test case to sample.yaml

Add a basic Robot Framework test suite utilizing the robot harness
runnable in Renode.

Signed-off-by: Michał Szprejda <mszprejda@antmicro.com>
Signed-off-by: Mateusz Hołenko <mholenko@antmicro.com>

mszprejda committed:
Twister: Add integration with renode-test

Add support for calling the `renode-test` command from west and twister.
Enable running Robot Framework tests suites in Renode.

Signed-off-by: Michał Szprejda <mszprejda@antmicro.com>
Signed-off-by: Mateusz Hołenko <mholenko@antmicro.com>

galak reviewed (dismissed):
Why does renode need a new way of describing tests in `sample.yaml`.  This works for both HW and qemu.  Should figure out how to make it work with renode as well.

PiotrZierhoffer commented:
The idea here is to be able to have more interactive tests.

Shell, with input, is just an example. But with these tests we'll be able to verify reactions to button presses, sensor input and more. This is not possible with currently available tests

nashif commented:
Thank you, this is great.

We were just talking about enhacing pytest support in zephyr to allow this type of testing and it was demonstrated based on the functionality the was added in twsiterv2.

Beside interactive testing, can you list other potential value we get with Robot? AFAIK this will also allow multinode testing, correct?

> Why does renode need a new way of describing tests in `sample.yaml`. This works for both HW and qemu. Should figure out how to make it work with renode as well.

I guess my question is the opposite of this one, what would it take to make everything work with robot and not limit it to renode...

nashif commented on doc/develop/test/twister.rst:
```diff
@@ -410,13 +410,16 @@ harness: <string>
     - test
     - console
     - pytest
+    - renode_test
```
can we put something generic here, like robot, that initially will be limited to renode but can be expanded to other simulators, and real hardware?

mateusz-holenko replied:
renamed to `robot` as suggested

PiotrZierhoffer commented:
> Beside interactive testing, can you list other potential value we get with Robot? AFAIK this will also allow multinode testing, correct?

Yes, multinode testing is also an option. Robot can be used with Renode to analyze many different things that are not directly accessible from software: GPIO status, network packet content, detailed information like executed function calls etc.

Robot itself is trivial to integrate with Python scripting, so we can have helper functions defining more complex logic, using all Python libraries etc.

> 
> I guess my question is the opposite of this one, what would it take to make everything work with robot and not limit it to renode...

This is actually quite possible and there are use cases where same tests are running against HW and Renode. It boils down to defining specific keywords and providing the implementation as a parameter to Robot call. By default we provide Renode keywords, and it's of course not super trivial to reimplement all functionalities to work on hw (e.g. testing of LED status or providing input - it has to be specific to the way we connect to hw and capabilities we have), but selecting the implementation is a matter of command line switch.

mateusz-holenko committed:
hifive_unleashed: Add RENODE_UART variable to board.cmake

Set RENODE_UART variable describing which UART device
should be used when running Robot Framework test on the platform.

Signed-off-by: Mateusz Hołenko <mholenko@antmicro.com>

mateusz-holenko commented:
Addressed remarks, fixed minor issues spotted on CI and rebased.

hakehuang commented:
@mateusz-holenko , I have a qeustion here, if we use renode for frdm_k64f, and the process is build frdm_k64f image, and run in renode envrionment. so in this case, renode is like a runner to us. If we can move the renode test into a runner for west, then we do not need introduce a robot system. just reuse all existing ones. but add a renode runner for supporting boards. what do you think?

PiotrZierhoffer commented:
@hakehuang it's a matter of a completely different flow. Robot allows you to interact with the board from the outside, inspecting its state and setting inputs.

With pure west-level integration we'd be able to return pass/fail, but it wouldn't really match current testing flow as well, where we usually wait for some string to appear on serial. This is not exactly what we want to achieve here.

Test integration gives us better control over test filtering (we only want to run these where Renode is supported). I also believe it makes more sense from the abstraction point of view.

Did I understand your question correctly?

hakehuang commented:
> @hakehuang it's a matter of a completely different flow. Robot allows you to interact with the board from the outside, inspecting its state and setting inputs.
> 
> With pure west-level integration we'd be able to return pass/fail, but it wouldn't really match current testing flow as well, where we usually wait for some string to appear on serial. This is not exactly what we want to achieve here.
> 
> Test integration gives us better control over test filtering (we only want to run these where Renode is supported). I also believe it makes more sense from the abstraction point of view.
> 
> Did I understand your question correctly?

@PiotrZierhoffer , thanks for the reply. my qeustion has two parts:
1. Would this be better to integrate as a west runner, as if we implement frdm_k64f board in renode, we just need to direct the runner as renode?
2. Interactiving with irobot does not depends on renode, so this can be a separated feature. Can we convert automatially testcase.yml to irobot file? As in current irobot file, they are just rephase the same content in testcase.yaml and add the few standard steps. so the flow would be, if there is no addition irobot file provoide, we still can use robot way with the `-t run_irobot_test`, and a irobot file for given testcase existing will override the testcase.yaml for irobot otpion.

mateusz-holenko commented:
CI detected problems with running the shell sample on one of the platforms.

We've confirmed it on the hardware (Renode+Robot did a good job) and are now working on a fix in the UART driver for MiV.

PerMac commented:
Can you please add a link to documentation, where the existing keywords (like `Write Line To Uart`) are listed/defined? It was not clear to me when I looked at the PR and during the demo you've explained that those are interfaces exposed by renode to robot, i.e. they are implemented on the renode side.  I believe it will be useful for users to know what keywords can already be used in their tests.

mateusz-holenko commented:
@PerMac The documentation on writing and running Robot Framework tests in Renode can be found in: https://renode.readthedocs.io/en/latest/introduction/testing.html. It provides a list of the most commonly used keywords together with links to the source code where those are defined.

For the list of keywords provided by the Robot Framework itself, refer to: https://robotframework.org/robotframework/.

Please also remember that it's possible to extend the framework by adding new keywords expressed directly in robot test suite files, as an external python library or, like Renode does it, dynamically via XML-RPC.

PiotrZierhoffer commented:
@hakehuang 

> 1. Would this be better to integrate as a west runner, as if we implement frdm_k64f board in renode, we just need to direct the runner as renode?

I'm not entirely sure. One of our goals is to be able to verify if we should run the test at all. This is controlled by 3 factors:

- should the test work for the given platform (this is the current filtering mechanism)
- is the platform supported in Renode at all
- is the specific feature supported in Renode

We used cmake-level integration because each board supported in Renode requires specific parameters to be defined.

If you add your board support to Renode, you still need to provide these parameters, e.g. the script (see [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/riscv/hifive1/board.cmake#L4)).

What would be the benefit of having the runner defined?

> 2. Interactiving with irobot does not depends on renode, so this can be a separated feature. Can we convert automatially testcase.yml to irobot file? As in current irobot file, they are just rephase the same content in testcase.yaml and add the few standard steps. so the flow would be, if there is no addition irobot file provoide, we still can use robot way with the `-t run_irobot_test`, and a irobot file for given testcase existing will override the testcase.yaml for irobot otpion.

This is partially true. Robot has many more features than current testcase.yaml can express. For example we can verify the shape of the signal generated by a PWM, we can analyze network packets, and we can call Renode functionality directly. This will not be usable for HW testing (or will be difficult), but we use this integrations in non-public contexts in this way as well.

It should be possible to use Robot runner for current testcases, but I'm not sure if this should be part of this PR.

PerMac commented:
> @PerMac The documentation on writing and running Robot Framework tests in Renode can be found in: https://renode.readthedocs.io/en/latest/introduction/testing.html. It provides a list of the most commonly used keywords together with links to the source code where those are defined.
> 
> For the list of keywords provided by the Robot Framework itself, refer to: https://robotframework.org/robotframework/.
> 
> Please also remember that it's possible to extend the framework by adding new keywords expressed directly in robot test suite files, as an external python library or, like Renode does it, dynamically via XML-RPC.

Thanks. I think I wasn't clear. My suggestion is to add those links to twister documentation (with the brief description you provided here).

nashif commented:
trying things out, getting this:
```
gale:zephyr(renode_test_integration): less /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/handler.log
[0/1] cd /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot && /bin/sh -c \ \ \ \ if\ [\ -z\ $ROBOT_FILES\ ]\ \;\ \ \ \ then\ \ \ \ \ \ \ \ echo\ ''\;\ \ \ \ \ \ \ \ echo\ '---\ Error:\ Robot\ file\ path\ is\ required\ to\ run\ Robot\ tests\ in\ Renode.\ To\ provide\ the\ path\ please\ set\ the\ ROBOT_FILES\ variable.'\;\ \ \ \ \ \ \ \ echo\ '---\ To\ rerun\ the\ test\ with\ west\ execute:'\;\ \ \ \ \ \ \ \ echo\ '---\ ROBOT_FILES=\<FILES\>\ west\ build\ -p\ -b\ \<BOARD\>\ -s\ \<SOURCE_DIR\>\ -t\ run_renode_test'\;\ \ \ \ \ \ \ \ echo\ ''\;\ \ \ \ \ \ \ \ exit\ 1\;\ \ \ \ fi\; && /usr/bin/renode-test --variable ELF:@/home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/zephyr/zephyr.elf --variable RESC:@/home/nashif/zephyrproject/zephyr/boards/riscv/hifive1/support/hifive1.resc --variable UART:sysbus.uart0 --variable KEYWORDS:/home/nashif/zephyrproject/zephyr/tests/robot/common.robot --results-dir /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot /home/nashif/zephyrproject/zephyr/samples/hello_world/$ROBOT_FILES
Traceback (most recent call last):
  File "/opt/renode/tests/run_tests.py", line 16, in <module>
    import robot_tests_provider
  File "/opt/renode/tests/robot_tests_provider.py", line 185, in <module>
    class KeywordsFinder(robot.model.SuiteVisitor):
AttributeError: module 'robot' has no attribute 'model'
FAILED: zephyr/CMakeFiles/run_renode_test /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/zephyr/CMakeFiles/run_renode_test
cd /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot && /bin/sh -c \ \ \ \ if\ [\ -z\ $ROBOT_FILES\ ]\ \;\ \ \ \ then\ \ \ \ \ \ \ \ echo\ ''\;\ \ \ \ \ \ \ \ echo\ '---\ Error:\ Robot\ file\ path\ is\ required\ to\ run\ Robot\ tests\ in\ Renode.\ To\ provide\ the\ path\ please\ set\ the\ ROBOT_FILES\ variable.'\;\ \ \ \ \ \ \ \ echo\ '---\ To\ rerun\ the\ test\ with\ west\ execute:'\;\ \ \ \ \ \ \ \ echo\ '---\ ROBOT_FILES=\<FILES\>\ west\ build\ -p\ -b\ \<BOARD\>\ -s\ \<SOURCE_DIR\>\ -t\ run_renode_test'\;\ \ \ \ \ \ \ \ echo\ ''\;\ \ \ \ \ \ \ \ exit\ 1\;\ \ \ \ fi\; && /usr/bin/renode-test --variable ELF:@/home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/zephyr/zephyr.elf --variable RESC:@/home/nashif/zephyrproject/zephyr/boards/riscv/hifive1/support/hifive1.resc --variable UART:sysbus.uart0 --variable KEYWORDS:/home/nashif/zephyrproject/zephyr/tests/robot/common.robot --results-dir /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot /home/nashif/zephyrproject/zephyr/samples/hello_world/$ROBOT_FILES
ninja: build stopped: subcommand failed.
```

mateusz-holenko commented:
@nashif Do you have the Robot Framework python module (`robotframework`) installed locally? The error you pasted suggests you might have the `robot` package installed, but this is a completely different tool.

Renode comes with [requirements.txt](https://github.com/renode/renode/blob/master/tests/requirements.txt) that lists all python packages (together with their versions) used by robot tests (including `robotframework` - the Robot Framework core itself), so calling `pip install -r requirements.txt` would be the simplest option to install it.

mateusz-holenko commented:
> My suggestion is to add those links to twister documentation (with the brief description you provided here).

@PerMac We've extended the twister documentation with a section on writing and running Robot tests.

mateusz-holenko commented:
The MiV shell problem is fixed by https://github.com/zephyrproject-rtos/zephyr/pull/57666 and now all tests pass on the CI.

nashif commented:
I must be missing something, still no success:

```
gale:zephyr(renode_test_integration): less /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/handler.log
[0/1] cd /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot && /bin/sh -c \ \ \ \ if\ [\ -z\ $ROBOT_FILES\ ]\ \;\ \ \ \ then\ \ \ \ \ \ \ \ echo\ ''\;\ \ \ \ \ \ \ \ echo\ '---\ Error:\ Robot\ file\ path\ is\ required\ to\ run\ Robot\ tests\ in\ Renode.\ To\ provide\ the\ path\ please\ set\ the\ ROBOT_FILES\ variable.'\;\ \ \ \ \ \ \ \ echo\ '---\ To\ rerun\ the\ test\ with\ west\ execute:'\;\ \ \ \ \ \ \ \ echo\ '---\ ROBOT_FILES=\<FILES\>\ west\ build\ -p\ -b\ \<BOARD\>\ -s\ \<SOURCE_DIR\>\ -t\ run_renode_test'\;\ \ \ \ \ \ \ \ echo\ ''\;\ \ \ \ \ \ \ \ exit\ 1\;\ \ \ \ fi\; && /usr/bin/renode-test --variable ELF:@/home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/zephyr/zephyr.elf --variable RESC:@/home/nashif/zephyrproject/zephyr/boards/riscv/hifive1/support/hifive1.resc --variable UART:sysbus.uart0 --variable KEYWORDS:/home/nashif/zephyrproject/zephyr/tests/robot/common.robot --results-dir /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot /home/nashif/zephyrproject/zephyr/samples/hello_world/$ROBOT_FILES
Preparing suites
Started Renode instance on port 9999; pid 4056144
Starting suites
Running /home/nashif/zephyrproject/zephyr/samples/hello_world/sample.robot
+++++ Starting test 'sample.Should Print Hello Message'
+++++ Finished test 'sample.Should Print Hello Message' in 0.01 seconds with status ESC[91mfailedESC[0m
      ╔═
      ║ No keyword with name 'Execute Command' found.
      ╚═
Suite /home/nashif/zephyrproject/zephyr/samples/hello_world/sample.robot failed! in 0.03 seconds.
Cleaning up suites
Closing Renode pid 4056144
Aggregating all robot results
Output:  /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/robot_output.xml
Log:     /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/log.html
Report:  /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/report.html
Some tests failed :( See the list of failed tests below and logs for details!
Failed robot critical tests:
        1. Test Suite.Should Print Hello Message
------
FAILED: zephyr/CMakeFiles/run_renode_test /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/zephyr/CMakeFiles/run_renode_test
cd /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot && /bin/sh -c \ \ \ \ if\ [\ -z\ $ROBOT_FILES\ ]\ \;\ \ \ \ then\ \ \ \ \ \ \ \ echo\ ''\;\ \ \ \ \ \ \ \ echo\ '---\ Error:\ Robot\ file\ path\ is\ required\ to\ run\ Robot\ tests\ in\ Renode.\ To\ provide\ the\ path\ please\ set\ the\ ROBOT_FILES\ variable.'\;\ \ \ \ \ \ \ \ echo\ '---\ To\ rerun\ the\ test\ with\ west\ execute:'\;\ \ \ \ \ \ \ \ echo\ '---\ ROBOT_FILES=\<FILES\>\ west\ build\ -p\ -b\ \<BOARD\>\ -s\ \<SOURCE_DIR\>\ -t\ run_renode_test'\;\ \ \ \ \ \ \ \ echo\ ''\;\ \ \ \ \ \ \ \ exit\ 1\;\ \ \ \ fi\; && /usr/bin/renode-test --variable ELF:@/home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot/zephyr/zephyr.elf --variable RESC:@/home/nashif/zephyrproject/zephyr/boards/riscv/hifive1/support/hifive1.resc --variable UART:sysbus.uart0 --variable KEYWORDS:/home/nashif/zephyrproject/zephyr/tests/robot/common.robot --results-dir /home/nashif/zephyrproject/zephyr/twister-out/hifive1/samples/hello_world/sample.basic.helloworld.robot /home/nashif/zephyrproject/zephyr/samples/hello_world/$ROBOT_FILES
ninja: build stopped: subcommand failed.
```

nashif commented:
> @nashif what version of Renode do you use?

was using 1.13.2, moved to 1.13.3 and it works now, thanks.

PiotrZierhoffer commented:
@galak @nashif @hakehuang since the tests are green we'd be happy to push this forward. What do you think?

PiotrZierhoffer commented on samples/hello_world/sample.yaml:
```diff
@@ -6,11 +6,15 @@ common:
+      type: one_line
+      regex:
+        - "Hello World! (.*)"
+  sample.basic.helloworld.robot:
```
@hakehuang we were thinking about this and we believe there are a few arguments against this "implicit" approach:

1. This will be a significant change to the way we generate a list of tests in Twister - we definitely would like to be able to run all test cases that are available, including both "classic" tests and Robot tests
2. With the new support we will be able to run tests that are not really similar to the ones we already have. It will be difficult to decide what to do with harness parameters, for example.
3. It will be quite common to have multiple Robot files per application - possibly with some parameters required etc. I think it will be actually quite **rare** to have a situation like this, where the Robot file would resemble the current test 1:1. We treat the hello-world more as a usage example, rather than added value. If you'd be more comfortable with removing the hello-world test case completely, we can of course do that.

In general I believe this automated approach you propose is worth having a further discussion, but I am not sure if this should be the first step?

nashif commented on samples/hello_world/sample.robot:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
```
This is a hello world sample for Zephyr, not for Robot or twister. To show how Robot works with Twister etc I would have it go into its own sample, test instead of overloading the hello world sample we have.

mateusz-holenko replied:
We believe that the name of the robot file might be misleading here. It has been named after `sample.yaml`, but in fact it's a specific test tailored for the Hello World application and therefore could probably be renamed to `test.robot` or even `hello_world.robot`. Additionally it might not be fortunate to discuss the Hello World case at all as it introduces a reimplementation (and therefore duplication) of an already-existing console-harness-based test.
To avoid confusion we should probably just remove this test completely.

The shell example Robot test on the other hand, provides an added value as it contains an interactive scenario that has not been tested automatically up to this point. It uses a newly-proposed testing framework integration to verify an already-existing sample.

Having a separate sample showcasing different aspects of writing Robot tests is an interesting idea - we’ll think about a proper scenario that could be instructive to future users. As it will take a while, we believe it would make most sense to have a separate PR adding this sample.

nashif replied:
> To avoid confusion we should probably just remove this test completely.

I am fine with that. 
I meant we should add anything like this into the hello_world sample to show robot/renode feature, we need to do that somewhere, so shell is fine.

nashif commented:
can you please fix the compliance check?

hakehuang commented:
maybe the PR context need change as you remove the hello_world

mateusz-holenko commented:
@nashif Fixed, compliance checks pass now.

@hakehuang Updated the PR description.

nashif commented on scripts/schemas/twister/platform-schema.yaml:
```diff
@@ -24,7 +24,8 @@ mapping:
     type: str
-    enum: ["qemu", "simics", "xt-sim", "renode", "nsim", "mdb-nsim", "tsim", "armfvp", "native"]
+    enum: ["qemu", "simics", "xt-sim", "renode", "nsim", "mdb-nsim", "tsim", "armfvp", "native",
+           "robot"]
```
why is this needed?

mateusz-holenko replied:
Thanks for spotting it, this was a leftover from the initial implementation.

mateusz-holenko commented:
@hakehuang The Robot version is verified by Renode itself (not Zephyr) to make sure internal tests use compatible syntax and keywords.
In fact the upstream Renode already targets RF `6.0.2` (see https://github.com/renode/renode/commit/5a592252573239dd2d536ed4f80fb9f57248b54d) - we are now in a process of preparing a next release of Renode and will update it in the Zephyr CI image as a next step.

tejlmand commented:
a little late.

This should never have been made as a build target.
CMake and ninja / make forms the build system as is too limited for what is needed here.

Instead, this should have been made as a `west` extension command.
See also: 
https://github.com/zephyrproject-rtos/zephyr/pull/66193#pullrequestreview-1817795562
